psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,257 @@
+
+        self.function = pytorch_function_creator(mechanism.function, device, context)
+        self._context = context
+        self.value = None
+        self._llvm_value = None
+        self.afferents = []
+        self.efferents = []
+
+        self._target_mechanism = None
+
+    def add_efferent(self, efferent):
+        if not efferent in self.efferents:
+            self.efferents.append(efferent)
+
+    def add_afferent(self, afferent):
+        if not afferent in self.afferents:
+            self.afferents.append(afferent)
+
+
+    def collate_afferents(self):
+        """
+        Returns weight-multiplied sum of all afferent projections
+        """
+        return sum([proj.execute(proj.sender.value) for proj in self.afferents])
+
+    def execute(self, variable):
+        self.value = self.function(variable)
+        # print(f"{self} input: {variable}\n")
+        # print(f"{self} output: {self.value}\n")
+        return self.value
+
+    def execute_llvm(self, ctx, builder, state, params, mech_input, data):
+        mech_func = ctx.import_llvm_function(self._mechanism)
+
+        mech_param = builder.gep(params, [ctx.int32_ty(0),
+                                                  ctx.int32_ty(0),
+                                                  ctx.int32_ty(self._idx)])
+
+        mech_state = builder.gep(state, [ctx.int32_ty(0),
+                                            ctx.int32_ty(0),
+                                            ctx.int32_ty(self._idx)])
+
+        mech_output = builder.gep(data, [ctx.int32_ty(0),
+                                            ctx.int32_ty(0),
+                                            ctx.int32_ty(self._idx)])
+
+        builder.call(mech_func, [mech_param, mech_state,
+                                    mech_input, mech_output])
+        
+        for i in range(len(mech_input.type.pointee.elements[0])):
+            pnlvm.helpers.printf_float_array(builder, builder.gep(mech_input, [ctx.int32_ty(0), ctx.int32_ty(0), ctx.int32_ty(i)]), prefix=f"{self} input:\n", override_debug=False)
+        pnlvm.helpers.printf_float_array(builder, builder.gep(mech_output, [ctx.int32_ty(0), ctx.int32_ty(0)]), prefix=f"{self} output:\n", override_debug=False)
+
+        return mech_output
+
+    def execute_derivative_func_llvm(self, ctx, builder, mech_input):
+        derivative_func = ctx.import_llvm_function(
+                    bin_function_derivative_creator(ctx, self._mechanism, context=self._context).name)
+        return gen_inject_unary_function_call(ctx, builder, derivative_func, mech_input)
+
+    def get_afferent_idx(self, afferent_node):
+        return [proj.receiver for proj in self.afferents].index(afferent_node)
+
+    def get_efferent_idx(self, efferent_node):
+        return [proj.sender for proj in self.efferents].index(efferent_node)
+
+    def __repr__(self): 
+        return "PytorchWrapper for: "+self._mechanism.__repr__() 
+
+class PytorchProjection():
```

Comment:
It's confusing to name things Projection if they don't inherit from `Projection`